### DISCUSSION TITLE
failing at capf writing, help?

### DISCUSSION THREAD

**Discussion by anarcat** [minad/corfu#432] [Q&A]
Hi!

First, I apologize as this is a little off-topic: my problem is not specific to corfu, and happens even with plain `completion-at-point`. But I've tried to get help elsewhere (`#emacs` on irc.libera.chat) and no one could figure it out. I figured people here have a *lot* of experience with capf and might be able to help better.

I want to implement [Interwiki](http://meatballwiki.org/wiki/InterWiki) completion. The idea is that you type, say, `wp:Emacs` and that automatically completes to `https://en.wikipedia.org/wiki/Emacs`. Or, even better in `markdown-mode`, it would autocomplete to `[Emacs](https://en.wikipedia.org/wiki/Emacs)`.

At first, I looked at `abbrev` and `dabbrev`, but I don't think those can work: they provide a static `keyword -> expansion` map, and this is different: we need to interpolate! So in other word, I *could* make an `abbrev` to turn `wp` in `https://en.wikipedia.org/wiki/` but not `wp:Emacs` into `[Emacs](https://en.wikipedia.org/wiki/Emacs)`.

I have also considered templating systems. I'm using yasnippet, but couldn't figure that out either; it doesn't seem that flexible, and having a full interwiki map feels clunky (i could be wrong though!).

So I got to work and, following your lead, tried to write a capf as it seems to be the "standard" way of doing this.

So I wrote this, after much trial and error:

```elisp
(defvar interwiki-dict
  '(("wp" . "https://en.wikipedia.org/wiki/")
    ("tpawiki" . "https://gitlab.torproject.org/tpo/tpa/team/-/wikis/")))

(defun interwiki-mdwn-complete-at-point ()
  "Complete the prefix:foo into a [foo](example.com/foo) markup."
  (message "interwiki-mdwn-complete-at-point: thing %s" (thing-at-point 'filename))
  (let* ((bds (bounds-of-thing-at-point 'filename))
         (start (car-safe bds))
         (end (cdr-safe bds))
         (match (buffer-substring-no-properties start end))
         (hunk (split-string match ":"))
         (key (pop hunk))
         (value (pop hunk))
         (urlmap (assoc key interwiki-dict)))
    (when urlmap
      (let* ((fmt (list (format "[%s](%s%s)" value (cdr-safe urlmap) value))))
        (message
         "interwiki-mdwn-complete-at-point: start %s end %s match: %s hunk %s key %s value %s urlmap %s fmt %s"
         start end match hunk key value urlmap fmt)
        (list start end fmt)))))

(add-hook 'markdown-mode-hook
          (lambda () (add-to-list 'completion-at-point-functions 'interwiki-mdwn-complete-at-point)))
```

If I type `wp:foo` and call <kbd>M-x completion-at-point</kbd>, I see the `(message)` call go through, all the way to the last one, so it *is* matching. But somehow, the autocompletion doesn't work.

I've even tried a simpler version, like this:

```elisp
(defun interwiki-complete-at-point ()
  "Complete URLs from the interwiki map.

This could also be implemented as an abbrev table, but is kept
here as a proof-of-concept to show capf works.  This function can
turn wikipedia into https://www.wikipedia.org/wiki/, but the
above cannot turn wikipedia:foo into
[foo](https://www.wikipedia.org/wiki/foo)."
  ;;(message "interwiki-complete-at-point: thing %s" (thing-at-point 'filename))
  (let* ((bds (bounds-of-thing-at-point 'filename))
         (start (car bds))
         (end (cdr bds))
         (match (buffer-substring-no-properties start end))
         (urlmap (assoc match interwiki-dict))
         (url (cdr-safe urlmap)))
    ;;(message
    ;; "interwiki-complete-at-point: bds %s start %s end %s match %s urlmap %s url %s"
    ;; bds start end match urlmap url)
    (list start end (list url))))

(add-hook 'markdown-mode-hook
          (lambda () (add-to-list 'completion-at-point-functions 'interwiki-complete-at-point)))
```

Mysteriously, *that* works: if I type `wp` and hit <kbd>M-x completion-at-point</kbd>, it does turn it into `https://en.wikipedia.org/wiki/`. So I *think* I figured out the capf interface...

For what it's worth, when i type `completion-at-point` on `wp:foo`, `*Messages*` says this:

```
interwiki-mdwn-complete-at-point: thing wp:foo
interwiki-mdwn-complete-at-point: start 16 end 22 match: wp:foo hunk nil key wp value foo urlmap (wp . https://en.wikipedia.org/wiki/) fmt ([foo](https://en.wikipedia.org/wiki/foo))
```

But why on earth doesn't it work for the other one? What am I doing wrong? 

is there a better way to do this?

Thanks! :)

**Comment by minad** [comment]
Hi, I suggest that you take a look at my Cape package. It defines many different Capfs and can hopefully serve as reference.

**Reply by anarcat** [reply]
i did look at cape! it's not using a programming style i'm super comfortable with, so I had trouble using it as a reference... but i did try! it's actually one of the reasons I fell here for help in the end. :)
